packages/core/src/browser/dialogs.ts

```diff
@@ -401,6 +401,68 @@ export async function confirmExit(): Promise<boolean> {
     return safeToExit === true;
 }
 
+export class ConfirmDialogSaveProps extends ConfirmDialogProps {
+    readonly save: string;
+    performSave: () => void;
+}
+
+export class ConfirmDialogSave extends ConfirmDialog {
+
+    protected saveButton: HTMLButtonElement | undefined;
+    constructor(
+        @inject(ConfirmDialogSaveProps) protected override readonly props: ConfirmDialogSaveProps
+    ) {
+        super(props);
+        this.contentNode.appendChild(this.createMessageNode(this.props.msg));
+        // reorder buttons
+        this.controlPanel.childNodes.forEach(child => this.controlPanel.removeChild(child));
+        [this.acceptButton, this.closeButton].forEach(child => {
+            if (typeof child !== 'undefined') {
+                this.controlPanel.appendChild(child);
+            }
+        });
+        this.appendSaveButton(props.save).addEventListener('click', () => { props.performSave(); this.acceptButton?.click(); });
+    }
+
+    protected appendSaveButton(text: string = Dialog.OK): HTMLButtonElement {
+        this.saveButton = this.createButton(text);
+        this.controlPanel.appendChild(this.saveButton);
+        this.saveButton.classList.add('main');
+        return this.saveButton;
+    }
+
+}
+
+export async function confirmExitWithOrWithoutSaving(captionsToSave: string[], performSave: () => void): Promise<boolean> {
+    let msg: string | HTMLElement = nls.localize('theia/core/quitMessage', 'Any unsaved changes will be lost.');
+
+    if (typeof captionsToSave !== 'undefined') {
+        const div = document.createElement('div');
+        div.innerText = nls.localize('theia/core/quitMessageFileList', 'The following have not yet been saved') + ':';
+        const span = document.createElement('span');
+
+        span.appendChild(document.createElement('br'));
+        captionsToSave.forEach(cap => {
+            const b = document.createElement('b');
+            b.innerText = cap;
+            span.appendChild(b);
+            span.appendChild(document.createElement('br'));
+        });
+        span.appendChild(document.createElement('br'));
+        div.appendChild(span);
+        msg = div;
+    }
+
+    const safeToExit = await new ConfirmDialogSave({
+        title: nls.localizeByDefault('Are you sure you want to quit?'),
```

Comment:
A default localization exists and we can pass the count:

```suggestion
        title: nls.localizeByDefault('Do you want to save the changes to the following {0} files?', count),
```